evm/src/executor/abi.rs

```diff
@@ -680,6 +681,291 @@ pub static HARDHAT_CONSOLE_SELECTOR_PATCHES: Lazy<HashMap<Selector, Selector>> =
     ])
 });
 
+enum FormatSpec {
+    String,  // %s
+    Number,  // %d
+    Integer, // %i
+    Object,  // %o
+}
+
+// FormatValue specifies how a value type is to be formatted
+trait FormatValue: UIfmt {
+    // Formats a value according to the FormatSpec
+    fn fmt(&self, spec: FormatSpec) -> String;
```

Comment:
return a String here is fine, since we clone the anyways, so

`FormatSpec::String => self.clone(),` is nbd